### ISSUE TITLE
[Gap Analysis] Implement Lion Optimizer

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#315]
### User Story

> As a researcher, I want to use the latest, high-performance optimizers specifically designed for large transformer models, so that I can achieve faster convergence and potentially better results than traditional optimizers like Adam.

---

### **Problem: Lack of Modern Transformer-Specific Optimizers**

While AiDotNet has a comprehensive set of traditional optimizers, it currently lacks newer optimizers specifically designed for the unique challenges of training very large transformer models. The Lion optimizer is a prominent example that has shown superior performance and memory efficiency in this domain.

---

### **Phase 1: Implement the Lion Optimizer**

**Goal:** Add the Lion optimizer, known for its efficiency and performance in large model training.

#### **AC 1.1: Create `LionOptimizer.cs` (13 points)**
**Requirement:** Implement the Lion optimizer algorithm.

-   [ ] **File:** `src/Optimizers/LionOptimizer.cs`
-   [ ] **Class:** `public class LionOptimizer<T, TInput, TOutput> : OptimizerBase<T, TInput, TOutput>`
-   [ ] **Algorithm:** Implement the Lion algorithm as described in the paper "Symbolic Discovery of Optimization Algorithms" (Chen et al., 2023).
    -   [ ] **State:** Lion maintains only a momentum (`_m`) state, unlike Adam which has both momentum and variance.
    -   [ ] **Update Rule:** The update rule is based on the sign of the momentum, scaled by the learning rate.
    -   [ ] **Parameters:** `learningRate`, `beta1` (for momentum), `beta2` (for weight decay, if applied).
-   [ ] **Method:** Implement `public override Vector<T> UpdateParameters(Vector<T> currentParameters, Vector<T> gradients)`.

#### **AC 1.2: Unit Tests for `LionOptimizer` (8 points)**
**Requirement:** Verify the correctness and behavior of the Lion optimizer.

-   [ ] **File:** `tests/UnitTests/Optimizers/LionOptimizerTests.cs`
-   [ ] **Test Cases:**
    -   [ ] Test `UpdateParameters` with simple, known gradients and parameters, verifying the update rule.
    -   [ ] Test with different `beta1` and `beta2` values.
    -   [ ] Test for numerical stability.
    -   [ ] Compare parameter updates against a reference implementation (e.g., PyTorch's Lion) for a small, fixed scenario.

---

### **Definition of Done**

-   [ ] All checklist items are complete.
-   [ ] `LionOptimizer` is implemented and unit-tested.
-   [ ] The optimizer correctly applies the Lion update rule.
-   [ ] All new tests pass.
---

## ⚠️ CRITICAL ARCHITECTURAL REQUIREMENTS

**Before implementing this user story, you MUST review:**
- **📋 Full Requirements:** [`.github/USER_STORY_ARCHITECTURAL_REQUIREMENTS.md`](../.github/USER_STORY_ARCHITECTURAL_REQUIREMENTS.md)
- **📐 Project Rules:** [`.github/PROJECT_RULES.md`](../.github/PROJECT_RULES.md)

#### 1. INumericOperations<T> Usage (CRITICAL)
- [ ] Include `protected static readonly INumericOperations<T> NumOps = MathHelper.GetNumericOperations<T>();` in base class
- [ ] NEVER hardcode `double`, `float`, or specific numeric types - use generic `T`
- [ ] NEVER use `default(T)` - use `NumOps.Zero` instead
- [ ] Use `NumOps.Zero`, `NumOps.One`, `NumOps.FromDouble()` for values
- [ ] Use `NumOps.Add()`, `NumOps.Multiply()`, etc. for arithmetic
- [ ] Use `NumOps.LessThan()`, `NumOps.GreaterThan()`, etc. for comparisons

#### 2. Inheritance Pattern (REQUIRED)
- [ ] Create `I{FeatureName}.cs` in `src/Interfaces/` (root level, NOT subfolders)
- [ ] Create `{FeatureName}Base.cs` in `src/{FeatureArea}/` inheriting from interface
- [ ] Create concrete classes inheriting from Base class (NOT directly from interface)

#### 3. PredictionModelBuilder Integration (REQUIRED)
- [ ] Add private field: `private I{FeatureName}<T>? _{featureName};` to `PredictionModelBuilder.cs`
- [ ] Add Configure method taking ONLY interface (no parameters):
  ```csharp
  public IPredictionModelBuilder<T, TInput, TOutput> Configure{FeatureName}(I{FeatureName}<T> {featureName})
  {
      _{featureName} = {featureName};
      return this;
  }
  ```
- [ ] Use feature in `Build()` with default: `var {featureName} = _{featureName} ?? new Default{FeatureName}<T>();`
- [ ] Verify feature is ACTUALLY USED in execution flow

#### 4. Beginner-Friendly Defaults (REQUIRED)
- [ ] Constructor parameters with defaults from research/industry standards
- [ ] Document WHY each default was chosen (cite papers/standards)
- [ ] Validate parameters and throw `ArgumentException` for invalid values

#### 5. Property Initialization (CRITICAL)
- [ ] NEVER use `default!` operator
- [ ] String properties: `= string.Empty;`
- [ ] Collections: `= new List<T>();` or `= new Vector<T>(0);`
- [ ] Numeric properties: appropriate default or `NumOps.Zero`

#### 6. Class Organization (REQUIRED)
- [ ] One class/enum/interface per file
- [ ] ALL interfaces in `src/Interfaces/` (root level)
- [ ] Namespace mirrors folder structure (e.g., `src/Regularization/` → `namespace AiDotNet.Regularization`)

#### 7. Documentation (REQUIRED)
- [ ] XML documentation for all public members
- [ ] `<b>For Beginners:</b>` sections with analogies and examples
- [ ] Document all `<param>`, `<returns>`, `<exception>` tags
- [ ] Explain default value choices

#### 8. Testing (REQUIRED)
- [ ] Minimum 80% code coverage
- [ ] Test with multiple numeric types (double, float)
- [ ] Test default values are applied correctly
- [ ] Test edge cases and exceptions
- [ ] Integration tests for PredictionModelBuilder usage

---

**⚠️ Failure to follow these requirements will result in repeated implementation mistakes and PR rejections.**

**See full details:** [`.github/USER_STORY_ARCHITECTURAL_REQUIREMENTS.md`](../.github/USER_STORY_ARCHITECTURAL_REQUIREMENTS.md)
